toucan_connectors/common.py

```diff
@@ -100,11 +100,54 @@ def _flatten_rendered_nested_list(origin: list, rendered: list) -> list:
     return result
 
 
-def _render_query(query: dict | list[dict] | tuple | str, parameters: dict):
+def _render_query(
+    query: dict | list[dict] | tuple | str, parameters: dict | None, handle_errors: bool = False
+):
     """
     Render both jinja or %()s templates in query
     while keeping type of parameters
     """
+
+    def _handle_missing_params(
+        elt: dict | list[dict] | tuple | str, params: dict, handle_errors: bool
+    ):
+        """
+        Remove a dictionary key if its value has a missing parameter.
+        cf. https://bit.ly/2Ln6rcf
+        """
+
+        if isinstance(elt, dict):
+            e = {}
+            for k, v in elt.items():
+                if isinstance(v, str):
+                    missing_params = []
+                    matches = re.findall(RE_PARAM, v) + re.findall(RE_JINJA, v)
+
+                    for m in matches:
+                        try:
+                            _ = Template('{{ %s }}' % m, undefined=StrictUndefined).render(params)
+                        except Exception:
+                            if handle_errors:
+                                raise NonValidVariable(f'Non valid variable {m}')
+                            missing_params.append(m)
+                    if any(missing_params):
+                        continue
+                    else:
+                        e[k] = v
+                else:
+                    e[k] = _handle_missing_params(v, params, handle_errors)
+            return e
+        elif isinstance(elt, list):
+            # Not filtering out empty dicts here because they may have a meaning in some backends
+            return [_handle_missing_params(e, params, handle_errors) for e in elt]
+        else:
+            return elt
+
+    query = _handle_missing_params(query, parameters or {}, handle_errors)
+
```

Comment:
> This is just copy-pasted
what's the issue if it's copy/pasted

> comments, naming, and docstrings are "out-of-date" ?
Can I know which ones ?

> undefined parameter checks can be handled with StrictUndefined
Do you have any sample ?

Reply:
> please, can I know, what's the issue if it's copy/pasted

In the previous review, I suggested that this was not needed anymore and should be handled by `_render_query`. The problem is that instead of removing the useless code and adapting `_render_query`, it was simply moved around.

> Do you have any sample/indications for me ?

Just pass `undefined=StrictUndefined` when instantiation the jinja `Environment`: https://jinja.palletsprojects.com/en/3.1.x/api/#jinja2.Environment

Reply:
>  In the previous review, I suggested that this was not needed anymore and should be handled by _render_query. The problem is that instead of removing the useless code and adapting _render_query, it was simply moved around.

We still need a recursive function to check for missing params right ?


>Just pass undefined=StrictUndefined when instantiation the jinja Environment: https://jinja.palletsprojects.com/en/3.1.x/api/#jinja2.Environment

Is it what we want at the end ? or do you suggest me to change results from tests ? 🤔 
because after doing this change : 
```python
        if is_jinja_alone(query):
            env = NativeEnvironment(undefined=StrictUndefined)
        else:
            env = Environment(undefined=StrictUndefined)
```
here : https://github.com/ToucanToco/toucan-connectors/blob/e8f9e9c8cfc73b961df988a73a73fecc3be8f9b2/toucan_connectors/common.py#L171 this test for example : 
`tests/test_common.py::test_nosql_apply_parameters_to_query[query8-params8-expected8]`

Is giving me this error : 
```
AssertionError: assert {'domain': 'blah', 'country': {'$ne': Undefined}, 'city': 'Paris'} == {'domain': 'blah', 'country': {}, 'city': 'Paris'}
```

Reply:
> We still need a recursive function to check for missing params right ?

I don't think we should do jinja's job. Determining if a param is missing by checking the type of `res`. If it's `Undefined`, we can act accordingly (silently remove it or raise an exception).

Side note: I don't think we should ever be silent about missing params, it makes debugging hard and will have weird behaviours in some cases. @fspot thoughts ?

Reply:
>  Determining if a param is missing by checking the type of res

The type of `res` ? what is `res` please ? you mean `result` ?

> If it's Undefined, we can act accordingly (silently remove it or raise an exception).

is it not what's happening in the function ? 🤔 
Or am reading that wrong ? 
```python
filtered_dict = {}
for key, value in query.items():
    if isinstance(value, str):
        matches = re.findall(RE_PARAM, value) + re.findall(RE_JINJA, value)
        missing_params = []
        for match in matches:
            try:
                Template('{{ %s }}' % match, undefined=StrictUndefined).render(params)
            except Exception:
                if handle_errors:
                    raise NonValidVariable(f'Non valid variable {match}')
                missing_params.append(match)

        if missing_params:
            continue

    filtered_dict[key] = _remove_undefined_params_variables(
        value, params, handle_errors
    )
```

Reply:
Does this goes better with what we should have at the end ? 
https://github.com/ToucanToco/toucan-connectors/pull/1086/commits/e57db4c